src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -34,49 +35,88 @@ class CassandraJournal extends AsyncWriteJournal with CassandraRecovery with Cas
-        Future.successful(Failure[Unit](e))
+    // we need to preserve the order / size of this sequence even though we don't map
+    // AtomicWrites 1:1 with a C* insert
+    val serialized = messages.map(aw => Try { SerializedAtomicWrite(
+        aw.payload.head.persistenceId,
+        aw.payload.map(pr => Serialized(pr.sequenceNr, persistentToByteBuffer(pr))))
     })
+    val result = serialized.map(a => a.map(_ => ()))
 
-    Future.sequence(batchStatements)
-  }
+    val byPersistenceId = serialized.collect({ case Success(caw) => caw }).groupBy(_.persistenceId).values
+    val boundStatements = byPersistenceId.map(statementGroup)
 
-  private def statementGroup(atomicWrite: AtomicWrite): Try[Seq[BoundStatement]] = Try {
-    atomicWrite.payload.flatMap { m =>
-      val pnr = partitionNr(m.sequenceNr)
-      val write = preparedWriteMessage.bind(m.persistenceId, pnr: JLong, m.sequenceNr: JLong, persistentToByteBuffer(m))
-      if (partitionNew(m.sequenceNr)) Seq(preparedWriteHeader.bind(m.persistenceId, pnr: JLong), write) else Seq(write)
+    val batchStatements = boundStatements.map({ unit =>
+        executeBatch(batch => unit.foreach(batch.add))
+    })
+    val promise: Promise[Seq[Try[Unit]]] = Promise[Seq[Try[Unit]]]()
+
+    Future.sequence(batchStatements).onComplete {
+      case Success(_) => promise.complete(Success(result))
+      case Failure(e) => promise.failure(e)
     }
+
+    promise.future
   }
 
-  private def asyncDeleteMessages(messageIds: Seq[MessageId]): Future[Unit] = executeBatch { batch =>
-    messageIds.foreach { mid =>
-      val stmt =
-        preparedDeletePermanent.bind(mid.persistenceId, partitionNr(mid.sequenceNr): JLong, mid.sequenceNr: JLong)
-      batch.add(stmt)
+  private def statementGroup(atomicWrites: Seq[SerializedAtomicWrite]): Seq[BoundStatement] = {
+    val maxPnr = partitionNr(atomicWrites.last.payload.last.sequenceNr)
+    val firstSeq: JLong = atomicWrites.head.payload.head.sequenceNr
+    val minPnr: JLong = partitionNr(firstSeq)
+    val persistenceId: String = atomicWrites.head.persistenceId
+    val all = atomicWrites.flatMap(_.payload)
+
+    // reading assumes sequence numbers are in the right partition or partition + 1
+    // even if we did allow this it would perform terribly as large C* batches are not good
+    require(maxPnr - minPnr <= 1, "Do not support AtomicWrites that span 3 partitions. Keep AtomicWrites <= max partition size.")
+
+    val writes: Seq[BoundStatement] = all.map { m =>
+      preparedWriteMessage.bind(persistenceId, maxPnr: JLong, m.sequenceNr: JLong, m.serialized)
     }
+    // in case we skip an entire partition we want to make sure the empty partition has in in-use flag so scans
+    // keep going when they encounter it
+    if (partitionNew(firstSeq) && minPnr != maxPnr) writes :+ preparedWriteInUse.bind(persistenceId, minPnr)
+    else writes
+
   }
 
   def asyncDeleteMessagesTo(persistenceId: String, toSequenceNr: Long): Future[Unit] = {
```

Comment:
+1 I think a single deletedTo per persistenceId is a good idea, it'll allow us to complete the future very quickly and n

It adds the case where we can be shutdown after the delete has succeeded but before we have completed all the deletes but the only side effect of this is that we need to clean up deletes, replay on restart is still correct. 

The more I think about these cases the more I want to have a jepson like property based test framework that generates random actions out of:
- Single write
- Atomic write
- Snapshot
- Restart Actor
- Restart JVM
- Restart C* node / cut off network between JVM and C*